bitcoin/src/bip32.rs

```diff
@@ -256,94 +367,230 @@ impl serde::Serialize for ChildNumber {
         u32::from(*self).serialize(serializer)
     }
 }
+/// normal_derivation_path module with inner methods
+pub mod normal_derivation_path {
+    use super::*;
 
-/// Trait that allows possibly failable conversion from a type into a
-/// derivation path
-pub trait IntoDerivationPath {
-    /// Converts a given type into a [`DerivationPath`] with possible error
-    fn into_derivation_path(self) -> Result<DerivationPath, Error>;
-}
+    ///A DerivationPath that can only contains NormalChildNumbers (i.e. no hardened indexes)
+    #[derive(Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+    pub struct NormalDerivationPath(Vec<NormalChildNumber>);
 
-/// A BIP-32 derivation path.
-#[derive(Clone, PartialEq, Eq, Ord, PartialOrd, Hash)]
-pub struct DerivationPath(Vec<ChildNumber>);
+    #[cfg(feature = "serde")]
+    internals::serde_string_impl!(
+        NormalDerivationPath,
+        "a BIP-32 derivation path containing normal numbers only"
+    );
+
+    impl NormalDerivationPath {
+        /// The master(empty) path (i.e. the master key).
+        #[allow(dead_code)]
+        const MASTER: Self = NormalDerivationPath(Vec::new());
+
+        /// Create a new `NormalDerivationPath` from a vector of `NormalChildNumber`
+        pub fn new(vec: Vec<NormalChildNumber>) -> Self { NormalDerivationPath(vec) }
+        /// Returns length of the `NormalDerivationPath`
+        pub fn len(&self) -> usize { self.0.len() }
+        /// Returns a `DerivationPath` from the `NormalDerivationPath`
+        pub fn to_derivation_path(&self) -> DerivationPath {
+            self.into_iter().map(|n| n.to_number()).collect()
+        }
 
-#[cfg(feature = "serde")]
-internals::serde_string_impl!(DerivationPath, "a BIP-32 derivation path");
+        /// Returns `true` if the `NormalDerivationPath` is empty
+        pub fn is_empty(&self) -> bool { self.0.is_empty() }
 
-impl<I> Index<I> for DerivationPath
-where
-    Vec<ChildNumber>: Index<I>,
-{
-    type Output = <Vec<ChildNumber> as Index<I>>::Output;
+        /// Returns whether `NormalDerivationPath` represents master key (i.e. it's length
+        /// is empty). True for `m` path.
+        pub fn is_master(&self) -> bool { self.0.is_empty() }
 
-    #[inline]
-    fn index(&self, index: I) -> &Self::Output { &self.0[index] }
-}
+        /// Create a new [NormalDerivationPath] that is a child of this one.
+        pub fn child(&self, cn: NormalChildNumber) -> NormalDerivationPath {
```

Comment:
In b80658ed8c3955883284fe93699825cbed685a0d

I apologize if this has been debated in the resolved comments, but I think it would be better to have a `push` method with the goal of more-or-less copying the [std `PathBuf` API](https://doc.rust-lang.org/std/path/struct.PathBuf.html).

The name `child` I think is particularly confusing because a derivation path is not the "child" of its truncated self.

Reply:
Though I would like there to be a functional "consume `self` and return it extended" method. For some inexplicable reason std doesn't have such methods on `PathBuf` or `Vec` so we don't have an API model. But I'd suggest `fn with_child` for this.

Reply:
>The name child I think is particularly confusing because a derivation path is not the "child" of its truncated self.

Actually i didnt get the point of this function... It does the same thing that `extend()` and the new function that i added rn `new_extended_from`
```Rust
   pub fn new_extended_from<T: IntoIterator<Item = ChildNumber>>(
            &self,
            path: T,
        ) -> DerivationPath {
            let mut ret = self.clone();
            ret.extend_from(path);
            ret
        }
```
You can have the same functionality of `child()` or `into_child()` with
```Rust
derivationpath.extend_from([ChildNumber]);
derivationpath.new_extended_from([ChildNumber]); 
```
>Though I would like there to be a functional "consume self and return it extended" method. For some inexplicable reason std doesn't have such methods on PathBuf or Vec so we don't have an API model. But I'd suggest fn with_child for this.

`into_with_child()` and `with_child()` ?

Reply:
I like the name `with_child`.

> You can have the same functionality of `child()` or `into_child()` with

Yes, but that's not really that nice.